precompiles/gmp/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+			});
+		}
+
+		// TODO: Wormhole might have transfered unsupported tokens; we should handle this case
+		//       gracefully (maybe that's as simple as reverting)
+
+		// TODO:
+		let weight_limit: u64 = 1_000_000_000_000u64;
+
+		log::debug!(target: "gmp-precompile", "sending XCM via xtokens::transfer...");
+		let call: orml_xtokens::Call<Runtime> = match user_action {
+			VersionedUserAction::V1(action) => orml_xtokens::Call::<Runtime>::transfer {
+				currency_id,
+				amount,
+				dest: Box::new(VersionedMultiLocation::V3(action.destination)),
+				dest_weight_limit: WeightLimit::Limited(Weight::from_parts(
+					weight_limit,
+					DEFAULT_PROOF_SIZE,
+				)),
+			},
+		};
+
+		log::debug!(target: "gmp-precompile", "sending xcm {:?}", call);
+
+		// TODO: proper origin
+		let origin = Runtime::AddressMapping::into_account_id(handle.code_address());
+		RuntimeHelper::<Runtime>::try_dispatch(handle, Some(origin).into(), call).map_err(|e| {
+			log::debug!(target: "gmp-precompile", "error sending XCM: {:?}", e);
+			e
+		})?;
+
+		Ok(())
+	}
+
+	/// call the given contract / function selector and return its output. Returns Err if the EVM
+	/// exit reason is not Succeed.
+	fn call(
+		handle: &mut impl PrecompileHandle,
+		contract_address: H160,
+		call_data: Vec<u8>,
+	) -> EvmResult<Vec<u8>> {
+		let sub_context = Context {
+			caller: handle.code_address(),
+			address: contract_address,
+			apparent_value: U256::zero(),
+		};
+
+		log::debug!(
+			target: "gmp-precompile",
+			"calling {} ...", contract_address,
+		);
+
+		let (reason, output) = handle.call(
+			contract_address,
+			None,
+			call_data,
+			handle.gas_limit(), // TODO
+			false,
+			&sub_context,
+		);
+
+		ensure_exit_reason_success(reason, &output[..])?;
+
+		Ok(output)
+	}
+}
+
+fn ensure_exit_reason_success(reason: ExitReason, output: &[u8]) -> EvmResult<()> {
+	log::trace!(target: "gmp-precompile", "reason: {:?}", reason);
+	log::trace!(target: "gmp-precompile", "output: {:x?}", output);
+
+	match reason {
+		ExitReason::Fatal(exit_status) => Err(PrecompileFailure::Fatal { exit_status }),
+		ExitReason::Revert(exit_status) => Err(PrecompileFailure::Revert {
+			exit_status,
+			output: output.into(),
+		}),
+		ExitReason::Error(exit_status) => Err(PrecompileFailure::Error { exit_status }),
+		ExitReason::Succeed(_) => Ok(()),
+	}
+}
+
+/// We use pallet storage in our precompile by implementing a StorageInstance for each item we need
+/// to store.
+/// twox_128("gmp") => 0xb7f047395bba5df0367b45771c00de50
+/// twox_128("CoreAddress") => 0x59ff23ff65cc809711800d9d04e4b14c
+/// twox_128("BridgeAddress") => 0xc1586bde54b249fb7f521faf831ade45
+mod storage {
+	use super::*;
+	use frame_support::{storage::types::StorageValue, traits::StorageInstance};
+
+	// storage for the core contract
+	pub struct CoreAddressStorageInstance;
+	impl StorageInstance for CoreAddressStorageInstance {
+		const STORAGE_PREFIX: &'static str = "CoreAddress";
```

Comment:
Theses storage items are too specific.
If we want the GMP precompile to be generic over any GMP protocol, replace them by a storage map with a protocol enum as a key and BoundedVec as a value, then decode the value according to the protocol type.

Reply:
Yes, we'll need to modify this a lot when we start supporting other protocols. We also need a way to make root be able to set these values, which might mean a new pallet.

Reply:
Resolving this for now, but happy to talk about it more.

Reply:
@notlesh at least please use only one storage item for wormhole data instead of two. 
otherwise you double the reads/writes for nothing.

Reply:
Also, we don't need a pallet, root can already use system.setStorage